R/retry.R

```diff
@@ -25,32 +33,55 @@
 #' RETRY("GET", "http://httpbin.org/status/200")
 #' # Never succeeds
 #' RETRY("GET", "http://httpbin.org/status/500")
+#' \donttest{
+#' # Invalid hostname generates curl error condition and is retried but eventually
+#' # raises an error condition.
+#' RETRY("GET", "http://invalidhostname/")
+#' }
 RETRY <- function(verb, url = NULL, config = list(), ...,
                   body = NULL, encode = c("multipart", "form", "json", "raw"),
                   times = 3, pause_base = 1, pause_cap = 60,
-                  handle = NULL, quiet = FALSE) {
+                  handle = NULL, quiet = FALSE, terminate_on = NULL) {
   stopifnot(is.numeric(times), length(times) == 1L)
   stopifnot(is.numeric(pause_base), length(pause_base) == 1L)
   stopifnot(is.numeric(pause_cap), length(pause_cap) == 1L)
+  stopifnot(is.integer(terminate_on) || is.numeric(terminate_on) || is.null(terminate_on))
 
   hu <- handle_url(handle, url, ...)
   req <- request_build(verb, hu$url, body_config(body, match.arg(encode)), config, ...)
-  resp <- request_perform(req, hu$handle$handle)
+  resp <- tryCatch(request_perform(req, hu$handle$handle), error = function(e) e)
 
   i <- 1
-  while (i < times && http_error(resp)) {
-    backoff_full_jitter(i, status_code(resp), pause_base, pause_cap, quiet = quiet)
+  while (!retry_should_stop(i, times, resp, terminate_on)) {
+    backoff_full_jitter(i, resp, pause_base, pause_cap, quiet = quiet)
 
     i <- i + 1
-    resp <- request_perform(req, hu$handle$handle)
+    resp <- tryCatch(request_perform(req, hu$handle$handle), error = function(e) e)
   }
 
-  resp
+  if ("error" %in% class(resp)) {
+    stop(resp)
+  } else {
+    return(resp)
+  }
+}
+
+retry_should_stop <- function(i, times, resp, terminate_on) {
+  if (i >= times) {
+    return(TRUE)
+  } else if ("error" %in% class(resp)) {
+    return(FALSE)
+  } else if (!is.null(terminate_on)) {
+    return(status_code(resp) %in% terminate_on)
+  } else {
+    return(!http_error(resp))
+  }
 }
 
-backoff_full_jitter <- function(i, status, pause_base = 1, pause_cap = 60, quiet = FALSE) {
+backoff_full_jitter <- function(i, resp, pause_base = 1, pause_cap = 60, quiet = FALSE) {
   length <- ceiling(stats::runif(1, max = min(pause_cap, pause_base * (2 ^ i))))
   if (!quiet) {
+    status <- if ("error" %in% class(resp)) as.character(resp) else status_code(resp)
```

Comment:
I really think it's useful to output what the encountered error was, otherwise the information on what errors happened during retries will be completely lost. IMHO outputting this information will be invaluable to httr users to debug transient errors they are encountering. i.e., being able to tell a DNS resolution failure from a failed connection attempt:

```
> RETRY("GET", "http://invalidhostname/")
Request failed [Error in curl::curl_fetch_memory(url, handle = handle): Couldn't resolve host name
]. Retrying in 1 seconds...
```

Any alternative ways to output this, perhaps?